### ISSUE TITLE
[java] jit code of collection with abstract class deserialize error

### ISSUE THREAD

**Issue by MrChang0** [apache/incubator-fury#1640]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-fury/issues) and found no similar issues.

### Version

master

### Component(s)

Java

### Minimal reproduce step

suche code
```java
  @Data
  abstract static class Foo {
      private int f1;
  }

  static class Foo1 extends Foo {}

    @Data
    static class CollectionAbstractTest {
      private List<Foo> foos;
    }

    @Test
    public void testCollectionAbstractCodegen() {
        Fury fury =
                Fury.builder()
                        .withCodegen(true)
                        .requireClassRegistration(false)
                        .build();
        CollectionAbstractTest test = new CollectionAbstractTest();
        test.foos = new ArrayList<>(ImmutableList.of(new Foo1(), new Foo1()));
        CollectionAbstractTest object = serDe(fury, test);
```
it create `sameElementClassWrite` function as
```java
  private void sameElementClassWrite(int value0, MemoryBuffer memoryBuffer1, java.util.List list2, boolean value1) {
      for (int i = 0; i < value0; i+=1) {
        Object object = list2.get(i);
        if (value1) {
            if ((object == null)) {
                memoryBuffer1.writeByte(((byte)-3));
            } else {
                memoryBuffer1.writeByte(((byte)0));
                fooClassInfoHolder.getSerializer().write(memoryBuffer1, object);
            }
        } else {
            fooClassInfoHolder.getSerializer().write(memoryBuffer1, object);
        }
      }
  }
```
casuse of it don't know abstract class will be same class or not. so it take a deseralize error.

### What did you expect to see?

i want it will be could deserialize right.

maybe it can generate code withou `serializer`, the `sameElementClassWrite` mothed  just like 
```java
      if (((value2 & 4) != 4)) {
          throw new RuntimeExcption();
      } else {
          serializer2 = classResolver.readClassInfo(memoryBuffer5, foo1ClassInfoHolder).getSerializer();
      }
```

### What did you see instead?

deserialize error.

### Anything Else?

btw. it have to iterator all data for check collection is same class. mostly time we won't take different class in one collection unless it like `Set<Object>`. maybe it can reduce array traversal.

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

**Comment by chaokunyang** [comment]
Hi @MrChang0 , thanks for reporting this bug, I can reproduce it locally. I fixed it in https://github.com/apache/incubator-fury/pull/1641. Could you take a look at it?

**Comment by MrChang0** [comment]
> Hi @MrChang0 , thanks for reporting this bug, I can reproduce it locally. I fixed it in #1641. Could you take a look at it?

thanks, i think it fixed my question.

**Comment by chaokunyang** [comment]
Turns out the inlineable Expression implemented before has some issues, I reverted the previous fix and  fixed it in a more low-level manner in https://github.com/apache/incubator-fury/pull/1641.. I believe this fix will be better and will resolve more broad range, could you test it again?
